### PR TITLE
Don't emit warning when not using sbt update

### LINKED ISSUE #174: recommend move to coursier
we should detect when people are using coursier https://github.com/ensime/ensime-server/pull/1312 and not spit out the standard sbt tickets to vote on.

Also, we should encourage people to move to coursier as well as voting.

### PR DESCRIPTION
**Pull request by jodersky**
Checks which update mechanism is used and only emits a warning when the default `sbt update` task is referenced.

Closes #174

### CODE REVIEW

**Comment by fommil** [comment]
Cool,


definingSetting doesn't seem to be used.

Let's lay the blame with SBT instead of ENSIME in the message, and remove the link to the upstream vote. Just telling people about coursier is enough.

I was actually just going to use reflection to detect if coursier was available, would have been a one liner :-) are you sure this is more reliable? It doesn't check for certain that the user is actually using coursier.

File: src/main/scala/EnsimePlugin.scala [hunk 1]
```diff
@@ -118,11 +118,30 @@ object EnsimePlugin extends AutoPlugin {
     EnsimeKeys.disableSourceMonitoring := false,
     EnsimeKeys.disableClassMonitoring := false,
     EnsimeKeys.megaUpdate <<= Keys.state.flatMap { implicit s =>
+
+      def checkUpdateMechanism(): Unit = {
+        val extracted = Project.extract(s)
+        val structure = extracted.structure
+
+        val settings = update.definingSettings(s, structure.allProjectRefs)
```
Comment (jodersky) [inline src/main/scala/EnsimePlugin.scala]: that's where `definingSettings` is used

**Comment by jodersky** [comment]
My initial thought was to just use something like `structure.allProjectRefs.flatMap(_.autoPlugins).contains(CoursierPlugin)`, however this would only check for coursier which is only a single implementation that fixes the underlying problem: detecting when sbt's update mechanism is actually used an report a message. (Btw, should this message be an info or go as far as being a warning?) This way, any custom update methods remove the message.

Let me know if you think it's too over-engineered, I can kind of see you point ;)

Finally, what message should I print? Doesn't "ENSIME is using the default sbt dependency update mechanism, which might take a long time" make the situation clear?

File: src/main/scala/EnsimePlugin.scala (hunk 1 above)
Comment (fommil) [inline src/main/scala/EnsimePlugin.scala]: heh, I should know better than to read code on a phone.

**Comment by fommil** [comment]
1. how about "SBT is using the default dependency update mechanism, ..."
2. give the getClass thing a try and see if it works :smile: I'd definitely prefer a simple one-liner to inspection of the structure.

**Comment by jodersky** [comment]
I don't quite see how to find the implementing task with reflection? Can you elaborate on what you had in mind?

Btw, I the structure inspection code is adapted from sbt https://github.com/sbt/sbt/blob/1.0.x/main/src/main/scala/sbt/Project.scala#L457-L506

**Comment by fommil** [comment]
Just try to load a class that you know is provided by coursier. It fails, show the warning to the user. It should be a simple one liner.

**Comment by jodersky** [comment]
OK I understand what you mean, in that case however I can just fallback to
my first idea and check if the plugin is present in any project's
autoplugins. The problem with just loading a class is that I could get a
false positive if coursier is actually on the classpath (defined in
plugins.sbt) but is disabled via `disablePlugin()`

**Comment by fommil** [comment]
I think you're overthinking it :smile: I always prefer a simple solution that works 99% of the time instead of something complex that handles a corner case nobody has seen yet. Merging because this is an improvement over the current situation, thanks!

**Comment by jodersky** [comment]
Thanks! although I was going to rework this into something simpler
I think we're misunderstanding each other, my other solution would also
have equated to a one-liner ;)

```
structure.allProjects.flatMap(_.autoPlugins).map(_.getClass.getName).contains("coursier")
```
I'll submit a new PR with the smaller fix if you like, I agree it's more
comprehensible, less code and covers 99% of all use-cases.

**Comment by fommil** [comment]
oh! Yes, that'd be good :smile: I have a small PR to do the rewording, much easier than going back and forth here. I'll merge that quickly so you can rebase on top of it for this change.

**Pull request merged**